### PR TITLE
Add Spicetify completions

### PR DESCRIPTION
**Pull request by Bergbok**
First time writing clink completions. Sorry if my approach is cursed, it works tho.

The only thing that doesn't work perfectly is chaining spicetify config arguments for eg:
`spicetify config inject_css 0 song_page 1 extensions myextensions.js`

### CODE REVIEW

File: completions/spicetify.lua
Comment (chrisant996) [inline completions/spicetify.lua:17]: You'll want `"2>nul spicetify.exe path -e"` so that any errors are not printed to the screen (which would garble the input line display).

The same suggestion applies to all instances of `io.popen` in this file.

File: completions/spicetify.lua
Comment (chrisant996) [inline completions/spicetify.lua:43]: Regarding this:

> The only thing that doesn't work perfectly is chaining spicetify config arguments for eg:
> `spicetify config inject_css 0 song_page 1 extensions myextensions.js`

Would you like for that work?

You could make it work by changing `create_arg` like this:

```lua
local one_arg = clink.argmatcher():addarg()

local function create_arg(name)
    return clink.argmatcher():addarg(name .. one_arg)
end
```

Then the `:loop(1)` in `config_parser` will have the desired effect.  The reason it doesn't is because of the `:nofiles()` above, which terminates the parser returned by `create_arg` so that it's dead-ended and all further parsing is blocked.

**Review by chrisant996 — CHANGES_REQUESTED** [review]
This looks quite fine, and especially for a first argmatcher.  Congrats!

I added two bits of feedback that should be quick and easy to apply, and which I think should help more fully achieve what you want.

File: completions/spicetify.lua
Comment (Bergbok) [reply completions/spicetify.lua]: Ah okay cool, changed it.

File: completions/spicetify.lua
Comment (Bergbok) [reply completions/spicetify.lua]: That worked perfectly.

Thanks for the explanation, makes sense.

**Comment by Bergbok** [comment]
Updated it to a state I'm satisfied with, should be ready to merge.

File: completions/spicetify.lua
Comment (chrisant996) [inline completions/spicetify.lua:69]: The `get_app_names(false)` here executes `spicetify.exe` when this Lua script is loaded, and then never again.  And the `get_extension_names(false)` executes it a second time when this Lua script is loaded.

It would be better to execute `spicetify.exe` when completion is performed.

To do that, pass only a function name here:
```lua
    :addarg("root", get_app_names)
```

If you want to pass `false` or `true`, then you need an intermediate function:

```lua
local function get_app_names_false()
    return get_app_names(false)
end

local function get_app_names_true()
    return get_app_names(true)
end

local path_a_flag_parser = clink.argmatcher()
    :addarg("root", get_app_names_false)
```

The Lua syntax `function_name` is just a reference to the function, so that the function can get called later, on demand.

The Lua syntax `function_name()` is an immediate call to the function, and then whatever the function returns is used.

It's a subtle detail about efficiency in argmatchers.

**Comment by chrisant996** [comment]
Yes, it looks great, with one caveat.  The `get_app_names(false)` and `get_extension_names(false)` syntax executes spicetify.exe twice when the Lua script is loaded.  See preceding comment for details, and a small tweak that improves efficiency (and avoids a delay while typing in the input line).

File: completions/spicetify.lua
Comment (Bergbok) [reply completions/spicetify.lua]: Ah cool, you learn something new everyday.

Once again, thanks for the explanation.

Just pushed a commit that fixes it.

**Comment by chrisant996** [comment]
Thanks!

**Comment by Bergbok** [comment]
Would you mind marking the review as completed? :)

WhatsApp Image 2024-04-09 at 12 11 29

**Comment by chrisant996** [comment]
> Would you mind marking the review as completed? :)
> 
> WhatsApp Image 2024-04-09 at 12 11 29

How are you determining that the review is not complete?

It looks like you're using the mobile github app.  The github web site doesn't show "❌ Reviews", and the pull request is marked as merged.

How do you want me to mark the review as completed?  Are you able to share a link to the review?  The screen shot of a summary tile about the review doesn't give me a way to do anything to the review, and it's already been merged.  The purple merge symbol indicates it's been merged successfully.  Doesn't that mean it's complete?

I can't find anything else to do.  You'll have to show me specifically what's needed, if you're experiencing some kind of problem (perhaps some contribution tracker isn't giving some kind of expected credit??).

**Review by Bergbok — COMMENTED** [review]
Done.

**Comment by Bergbok** [comment]
Ah I see, anyway thanks for having a look.

Might be a mobile app issue.

**Comment by chrisant996** [comment]
It isn't about marking it complete.  It's already marked complete.

The problem is that the "requested changes" review was merged without first marking it "approved".  But once a pull request is merged, github seems to have no way to interact with any reviews associated with the pull requests.  Comments from reviews are visible, but the review itself in inaccessible and can no longer be marked as "approved".

I didn't see any "approve" button in the mobile app, I only saw a "merge and squash" button.  I didn't imagine that completing it by merging would not approve it, nor that it would become stuck that way.  It appears to be a couple of loophole bugs in the github workflow for pull requests.

**Comment by chrisant996** [comment]
One last try; I'll add a new review from me, and then approve it.  Maybe that will let me also interact with the previous one, or maybe github will realize obviously I've approved it.

**Comment by Bergbok** [comment]
mhm

**Review by chrisant996 — COMMENTED** [review]
Adding a placeholder review to try to help github get into an accurate state.

**Comment by chrisant996** [comment]
Ahhh.  @vladimir-kotikov owns the repo, and I do not have sufficient administrator privileges to reopen a closed pull request.  That's why I'm blocked from approving the review after the merge has happened.

**Comment by chrisant996** [comment]
...but further reading suggests that it's impossible to reopen a _merged_ pull request.

This is indeed a loophole in github's workflow.  Once a PR is merged, it's impossible to mark any reviews associated with it as "approved".  But you're able to create new reviews, even.  Doh.

**Review by chrisant996 — APPROVED** [review]

**Comment by chrisant996** [comment]
AHA!  The web UI has no way.  But the `gh` tool is able to do it:  `gh pr review --approve 183`.  😎

"No power in the 'verse can stop me."
-- River Tam

**Comment by Bergbok** [comment]
Hahah nice job!!

Good to know DankNoted

**Pull request merged**
